File: sdk/typescript/src/state-map.ts
/**
 * Bidirectional state mappings between Bernstein task states and external trackers.
 *
 * @example
 * ```ts
 * import { jiraToBernstein, bernsteinToJira } from '@bernstein/sdk/state-map';
 *
 * const status = jiraToBernstein('In Progress'); // → 'in_progress'
 * const label  = bernsteinToJira('done');         // → 'Done'
 * ```
 */

import type { TaskStatus } from './models.js';

// ---------------------------------------------------------------------------
// Jira ↔ Bernstein
// ---------------------------------------------------------------------------

const JIRA_TO_BERNSTEIN: Record<string, TaskStatus> = {
  'backlog': 'open',
  'to do': 'open',
  'open': 'open',
  'new': 'open',
  'selected for development': 'open',
  'in progress': 'in_progress',
  'in review': 'in_progress',
  'under review': 'in_progress',
  'code review': 'in_progress',
  'blocked': 'blocked',
  'waiting': 'blocked',
  'on hold': 'blocked',
  'waiting for review': 'blocked',
  'done': 'done',
  'closed': 'done',
  'resolved': 'done',
  "won't do": 'cancelled',
  'wont do': 'cancelled',
  'cancelled': 'cancelled',
  'duplicate': 'cancelled',
};

const BERNSTEIN_TO_JIRA: Record<TaskStatus, string> = {
  open: 'To Do',
  claimed: 'In Progress',
  in_progress: 'In Progress',
  done: 'Done',
  failed: 'Done',
  blocked: 'Blocked',
  cancelled: "Won't Do",
  orphaned: 'To Do',
};

/** Project-specific overrides applied on top of the default mapping. */
const jiraOverrides: Record<string, TaskStatus> = {};
const bernsteinToJiraOverrides: Partial<Record<TaskStatus, string>> = {};

/**
 * Map a Jira issue status name (case-insensitive) to a {@link TaskStatus}.
 * Unknown statuses return `fallback` (default: `"open"`).
 */
export function jiraToBernstein(
  jiraStatus: string,
  fallback: TaskStatus = 'open',
): TaskStatus {
  const key = jiraStatus.toLowerCase().trim();
  return jiraOverrides[key] ?? JIRA_TO_BERNSTEIN[key] ?? fallback;
}

/**
 * Map a {@link TaskStatus} to a Jira status transition name.
 */
export function bernsteinToJira(status: TaskStatus): string {
  return bernsteinToJiraOverrides[status] ?? BERNSTEIN_TO_JIRA[status];
}

/**
 * Register a project-specific Jira status mapping that overrides the defaults.
 */
export function registerJiraMapping(
  jiraStatus: string,
  bernsteinStatus: TaskStatus,
): void {
  jiraOverrides[jiraStatus.toLowerCase().trim()] = bernsteinStatus;
}

/**
 * Override the default Jira target for a given Bernstein state.
 */
export function registerBernsteinToJiraMapping(
  bernsteinStatus: TaskStatus,
  jiraStatus: string,
): void {
  bernsteinToJiraOverrides[bernsteinStatus] = jiraStatus;
}

// ---------------------------------------------------------------------------
// Linear ↔ Bernstein
// ---------------------------------------------------------------------------

const LINEAR_TO_BERNSTEIN: Record<string, TaskStatus> = {
  triage: 'open',
  backlog: 'open',
  unstarted: 'open',
  todo: 'open',
  started: 'in_progress',
  'in progress': 'in_progress',
  'in review': 'in_progress',
  completed: 'done',
  done: 'done',
  cancelled: 'cancelled',
  canceled: 'cancelled',
  duplicate: 'cancelled',
  blocked: 'blocked',
  waiting: 'blocked',
};

const BERNSTEIN_TO_LINEAR: Record<TaskStatus, string> = {
  open: 'Todo',
  claimed: 'In Progress',
  in_progress: 'In Progress',
  done: 'Done',
  failed: 'Cancelled',
  blocked: 'Blocked',
  cancelled: 'Cancelled',
  orphaned: 'Todo',
};

const linearOverrides: Record<string, TaskStatus> = {};
const bernsteinToLinearOverrides: Partial<Record<TaskStatus, string>> = {};

/**
 * Map a Linear issue state name or type (case-insensitive) to a {@link TaskStatus}.
 */
export function linearToBernstein(
  linearState: string,
  fallback: TaskStatus = 'open',
): TaskStatus {
  const key = linearState.toLowerCase().trim();
  return linearOverrides[key] ?? LINEAR_TO_BERNSTEIN[key] ?? fallback;
}

/**
 * Map a {@link TaskStatus} to a Linear state name.
 */
export function bernsteinToLinear(status: TaskStatus): string {
  return bernsteinToLinearOverrides[status] ?? BERNSTEIN_TO_LINEAR[status];
}

/** Register a workspace-specific Linear state mapping. */
export function registerLinearMapping(
  linearState: string,
  bernsteinStatus: TaskStatus,
): void {
  linearOverrides[linearState.toLowerCase().trim()] = bernsteinStatus;
}

/** Override the default Linear target for a given Bernstein state. */
export function registerBernsteinToLinearMapping(
  bernsteinStatus: TaskStatus,
  linearState: string,
): void {
  bernsteinToLinearOverrides[bernsteinStatus] = linearState;
}

File: sdk/python/src/bernstein_sdk/state_map.py
"""Bidirectional state mappings between Bernstein task states and external trackers.

Usage::

    from bernstein_sdk.state_map import JiraToBernstein, BernsteinToJira

    b_status = JiraToBernstein.map("In Progress")   # → TaskStatus.IN_PROGRESS
    jira_status = BernsteinToJira.map(TaskStatus.DONE)  # → "Done"
"""

from __future__ import annotations

from bernstein_sdk.models import TaskStatus

# ---------------------------------------------------------------------------
# Jira ↔ Bernstein
# ---------------------------------------------------------------------------

# These are the default Jira status names shipped with the standard
# "Scrum" and "Kanban" project templates.  Projects with custom workflows
# will need to override mappings via JiraToBernstein.register().
_JIRA_TO_BERNSTEIN: dict[str, TaskStatus] = {
    # Backlog / todo
    "backlog": TaskStatus.OPEN,
    "to do": TaskStatus.OPEN,
    "open": TaskStatus.OPEN,
    "new": TaskStatus.OPEN,
    "selected for development": TaskStatus.OPEN,
    # Active
    "in progress": TaskStatus.IN_PROGRESS,
    "in review": TaskStatus.IN_PROGRESS,
    "under review": TaskStatus.IN_PROGRESS,
    "code review": TaskStatus.IN_PROGRESS,
    # Waiting
    "blocked": TaskStatus.BLOCKED,
    "waiting": TaskStatus.BLOCKED,
    "on hold": TaskStatus.BLOCKED,
    "waiting for review": TaskStatus.BLOCKED,
    # Terminal
    "done": TaskStatus.DONE,
    "closed": TaskStatus.DONE,
    "resolved": TaskStatus.DONE,
    "won't do": TaskStatus.CANCELLED,
    "wont do": TaskStatus.CANCELLED,
    "cancelled": TaskStatus.CANCELLED,
    "duplicate": TaskStatus.CANCELLED,
}

_BERNSTEIN_TO_JIRA: dict[TaskStatus, str] = {
    TaskStatus.OPEN: "To Do",
    TaskStatus.CLAIMED: "In Progress",
    TaskStatus.IN_PROGRESS: "In Progress",
    TaskStatus.DONE: "Done",
    TaskStatus.FAILED: "Done",  # fallback — Jira has no native "Failed"
    TaskStatus.BLOCKED: "Blocked",
    TaskStatus.CANCELLED: "Won't Do",
    TaskStatus.ORPHANED: "To Do",  # re-open for retry
}


class JiraToBernstein:
    """Map a Jira issue status name to a :class:`TaskStatus`.

    Case-insensitive.  Unknown statuses fall back to ``TaskStatus.OPEN``.

    Example::

        status = JiraToBernstein.map("In Progress")   # → TaskStatus.IN_PROGRESS
        status = JiraToBernstein.map("Custom Status") # → TaskStatus.OPEN (fallback)
    """

    _overrides: dict[str, TaskStatus] = {}

    @classmethod
    def map(
        cls, jira_status: str, fallback: TaskStatus = TaskStatus.OPEN
    ) -> TaskStatus:
        """Map *jira_status* (case-insensitive) to a :class:`TaskStatus`."""
        key = jira_status.lower().strip()
        return cls._overrides.get(key) or _JIRA_TO_BERNSTEIN.get(key, fallback)

    @classmethod
    def register(cls, jira_status: str, bernstein_status: TaskStatus) -> None:
        """Register a project-specific mapping that overrides the defaults.

        Args:
            jira_status: The Jira status label (case-insensitive).
            bernstein_status: The :class:`TaskStatus` to map to.
        """
        cls._overrides[jira_status.lower().strip()] = bernstein_status


class BernsteinToJira:
    """Map a :class:`TaskStatus` to a Jira status transition name.

    Example::

        label = BernsteinToJira.map(TaskStatus.DONE)  # → "Done"
    """

    _overrides: dict[TaskStatus, str] = {}

    @classmethod
    def map(cls, status: TaskStatus, fallback: str = "To Do") -> str:
        """Return the Jira status name for *status*."""
        return cls._overrides.get(status) or _BERNSTEIN_TO_JIRA.get(status, fallback)

    @classmethod
    def register(cls, bernstein_status: TaskStatus, jira_status: str) -> None:
        """Override the default Jira target for a given Bernstein state."""
        cls._overrides[bernstein_status] = jira_status


# ---------------------------------------------------------------------------
# Linear ↔ Bernstein
# ---------------------------------------------------------------------------

# Linear uses fixed state *types*: triage, backlog, unstarted, started,
# completed, cancelled.  Projects may also add custom state *names* within
# each type.  We map both type and common names.
_LINEAR_TO_BERNSTEIN: dict[str, TaskStatus] = {
    # Type-level
    "triage": TaskStatus.OPEN,
    "backlog": TaskStatus.OPEN,
    "unstarted": TaskStatus.OPEN,
    "todo": TaskStatus.OPEN,
    "started": TaskStatus.IN_PROGRESS,
    "in progress": TaskStatus.IN_PROGRESS,
    "in review": TaskStatus.IN_PROGRESS,
    "completed": TaskStatus.DONE,
    "done": TaskStatus.DONE,
    "cancelled": TaskStatus.CANCELLED,
    "canceled": TaskStatus.CANCELLED,
    "duplicate": TaskStatus.CANCELLED,
    "blocked": TaskStatus.BLOCKED,
    "waiting": TaskStatus.BLOCKED,
}

_BERNSTEIN_TO_LINEAR: dict[TaskStatus, str] = {
    TaskStatus.OPEN: "Todo",
    TaskStatus.CLAIMED: "In Progress",
    TaskStatus.IN_PROGRESS: "In Progress",
    TaskStatus.DONE: "Done",
    TaskStatus.FAILED: "Cancelled",  # Linear has no native "Failed"
    TaskStatus.BLOCKED: "Blocked",
    TaskStatus.CANCELLED: "Cancelled",
    TaskStatus.ORPHANED: "Todo",
}


class LinearToBernstein:
    """Map a Linear issue state name to a :class:`TaskStatus`.

    Matches against both the state *type* (``"started"``) and the state
    *name* (``"In Progress"``), case-insensitive.  Unknown values fall back
    to ``TaskStatus.OPEN``.

    Example::

        status = LinearToBernstein.map("In Progress")  # → TaskStatus.IN_PROGRESS
        status = LinearToBernstein.map("started")       # → TaskStatus.IN_PROGRESS
    """

    _overrides: dict[str, TaskStatus] = {}

    @classmethod
    def map(
        cls, linear_state: str, fallback: TaskStatus = TaskStatus.OPEN
    ) -> TaskStatus:
        """Map *linear_state* (case-insensitive) to a :class:`TaskStatus`."""
        key = linear_state.lower().strip()
        return cls._overrides.get(key) or _LINEAR_TO_BERNSTEIN.get(key, fallback)

    @classmethod
    def register(cls, linear_state: str, bernstein_status: TaskStatus) -> None:
        """Register a workspace-specific state mapping."""
        cls._overrides[linear_state.lower().strip()] = bernstein_status


class BernsteinToLinear:
    """Map a :class:`TaskStatus` to a Linear state name.

    Example::

        name = BernsteinToLinear.map(TaskStatus.DONE)  # → "Done"
    """

    _overrides: dict[TaskStatus, str] = {}

    @classmethod
    def map(cls, status: TaskStatus, fallback: str = "Todo") -> str:
        """Return the Linear state name for *status*."""
        return cls._overrides.get(status) or _BERNSTEIN_TO_LINEAR.get(status, fallback)

    @classmethod
    def register(cls, bernstein_status: TaskStatus, linear_state: str) -> None:
        """Override the default Linear target for a given Bernstein state."""
        cls._overrides[bernstein_status] = linear_state
